src/ember-app.js

```diff
@@ -119,16 +119,12 @@ class EmberApp {
       debug("module whitelisted; module=%s", whitelistedModule);
     });
 
-    return function(moduleName) {
-      if (whitelist.indexOf(moduleName) > -1) {
-        let nodeModulesPath = path.join(distPath, 'node_modules', moduleName);
+    return function(modulePath) {
+      let moduleName = modulePath.split('/')[0];
 
-        if (existsSync(nodeModulesPath)) {
-          return require(nodeModulesPath);
-        } else {
-          // If it's not on disk, assume it's a built-in node package
-          return require(moduleName);
-        }
+      if (whitelist.indexOf(moduleName) > -1) {
```

Comment:
As far as I can tell, FastBoot supports only dependencies to be whitelisted (eq. `abortcontroller-polyfill`), not a path to module in dependency (eq. `abortcontroller-polyfill/dist/cjs-ponyfill`), http://ember-fastboot.com/docs/user-guide#using-whitelisted-node-dependencies.

In `ember-fetch`, whitelisting of `abortcontroller-polyfill/dist/cjs-ponyfill` [here](https://github.com/ember-cli/ember-fetch/blob/master/package.json#L69), caused that the `abortcontroller-polyfill` dependency was not added (and installed) into build and then requiring module at runtime caused this error https://github.com/ember-cli/ember-fetch/issues/115.

This change allows to specify FastBoot dependency and then require any module from this dependency, so this should not be a breaking change.

Reply:
That ember-fetch issue is caused by `node_modules` not inside `dist/`. If you run the server at the path that has `node_modules` the module can be installed.

---

If your change is ~`whitelist.indexOf(moduleName)`~ `whitelist.indexOf(modulePath)`, it will not be a breaking change for any app/addon who's requiring a specific file within a module.

I'm not sure what the official convention for Fastboot is. If users should whitelist a package rather than any subset of its file/module, we can make changes and document, but it's still a breaking change for current whitelisting a module's certain file...

Reply:
> That ember-fetch issue is caused by node_modules not inside dist/. If you run the server at the path that has node_modules the module can be installed.

After `ember build` and before running server, it's required to `cd dist` and `npm install` to have FastBoot whitelisted dependencies available for running, which are part of `dist/package.json`, eq. like `fastboot-s3-downloader` does https://github.com/tomdale/fastboot-s3-downloader/blob/master/index.js#L110.

> I'm not sure what the official convention for Fastboot is. If users should whitelist a package rather than any subset of its file/module, we can make changes and document, but it's still a breaking change for current whitelisting a module's certain file...

I hope that there is no need to make changes, because it's already said [here](https://github.com/ember-fastboot/ember-cli-fastboot#whitelisting-packages), that `fastbootDependencies` contains packages (and Node.js built-ins), based on that I believe that this PR is not introducing any breaking change.

Reply:
Thanks for the explanation @bobisjan 

Yep the `dist/node_modules` are required, once it's done that issue won't happen.

Definition of "breaking change" is vague here, my argument above was referring to the actual implementation, yours was referring to the documentation. If it is considered wrong for addons/apps to implement based on the current fastboot I can totally accept, just want to point out it will break some addons/apps and they need to fix themselves.

And this PR also need to deal with the scoped package case in this line `let moduleName = modulePath.split('/')[0];`?

Reply:
> Yep the dist/node_modules are required, once it's done that issue won't happen.

🎉

> Definition of "breaking change" is vague here, my argument above was referring to the actual implementation, yours was referring to the documentation. If it is considered wrong for addons/apps to implement based on the current fastboot I can totally accept, just want to point out it will break some addons/apps and they need to fix themselves.

I see, the current implementation (without this PR) based on

1. https://github.com/ember-fastboot/ember-cli-fastboot/blob/master/lib/broccoli/fastboot-config.js#L86-L111
2. https://github.com/ember-fastboot/fastboot/blob/master/src/ember-app.js#L117-L136

requires for `FastBoot.require('abortcontroller-polyfill/dist/cjs-ponyfill')` to work to specify FastBoot dependencies like that

```js
dependecies: {
  'abortcontroller-polyfill': '^1.1.9'
},
fastbootDependencies: [
  'abortcontroller-polyfill' // (a)
  'abortcontroller-polyfill/dist/cjs-ponyfill' // (b)
]
```

- (a) ensures that package itself appears in `dist/package.json` (1) and will be installed in `dist/node_modules` (also puts it into `whitelist` and allows to require it),
- (b) puts path into `whitelist` and allows to require it (2).

With this PR the (b) case will become unnecessary and can be removed, but current behaviour in apps/addons should be preserved.

Looking at https://emberobserver.com/code-search?codeQuery=fastbootDependencies, there are `ember-fetch` a some `firebase` related addons, which whitelists without package name.

To be honest I'm curious how these addons can work in FastBoot, because IMO based on our application running on AWS using FastBoot App Server and finding https://github.com/simplabs/ember-simple-auth/issues/1690, https://github.com/ember-cli/ember-fetch/issues/115, apps/addons with just (b) and without (a) do not work in FastBoot App Server at all today regardless of this PR and need to be fixed anyway.

Maybe users don't use FastBoot App Server and only use `ember serve` 🤔? If it works (only having (b)) using `ember serve` then it would break their apps like you mention, will try and report back.

I would like to get feedback from @kratiahuja / @tomdale / @rwjblue / @ember-fastboot on that ☺️.

> And this PR also need to deal with the scoped package case in this line `let moduleName = modulePath.split('/')[0];`?

👍 Good catch, will update, thanks!